### ISSUE TITLE
findRelOffset seg-faults when tested in Cuda builds (reenabled in the CMake update branch)

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos-kernels#498]
This occurred while testing the CMake PR #491 with Cuda. 

Testing of `sparse_findRelOffset_*` was disabled in Cuda builds in the Makefile build system with a note that this is known to only run sequentially on host (same with trsv). The test is reenabled in the CMake PR.

The failure is because findRelOffset (in kokkos-kernels/src/sparse/KokkosSparse_findRelOffset.hpp) accepts Views with allocations on the GPU, however throughout the function all code executes on the host, trying to access memory from the GPU resulting in the seg faults. This happens to work in Trilinos because UVM is required, allowing host-access of the GPU allocation.

This directly impacts Tpetra and the desire to move away from requiring UVM.

First question is do we disable this test for the 3.0 release (preserving existing behavior), or address this in a more robust way now?

Adding @srajama1 @crtrott @mhoemmen @jjwilke

**Comment by ndellingwood** [comment]
I'm marking as a bug since proper safety to handle this use case appropriately is missing, though rewriting the routine to execute on device would be an enhancement.

**Comment by ndellingwood** [comment]
Looks like revision of this routine was considered in issue #10

**Comment by srajama1** [comment]
I don't envision this function being called anywhere outside Trilinos, however KK should not depend on UVM. Let us fix this for 3.0.

**Comment by mhoemmen** [comment]
@ndellingwood wrote:
> however throughout the function all code executes on the host,

Do you mean, throughout `findRelOffset` itself, or throughout whatever function is calling `findRelOffset`?  The `findRelOffset` function is marked as `KOKKOS_FUNCTION` and can certainly be called from either device or host.

**Comment by ndellingwood** [comment]
@mhoemmen I didn't phrase very clearly, sorry, let me try again:

> The `findRelOffset` function is marked as `KOKKOS_FUNCTION` and can certainly be called from either device or host. 

Yes, along with the possibility this routine can be called as a free function from the host with Views or pointers to device data (which is what the unit test was doing). The View case can be checked easily by checking if the data is accessible on host. The tougher check is when a pointer to device data is passed to the function and `findRelOffset` is called from the host, which occurs in the unit test and caused the seg fault with Cuda (the input indices only needs to have `operator[]` implemented).

**Comment by ndellingwood** [comment]
I had focused on the failing test case (calling function from host with data on device); when the function is called within a device kernel then UVM is clearly not required, but when called from the host with arguments pointing to device data UVM is required to avoid the seg fault. The UVM requirement I mentioned was based on assumption that the Cuda test is run within Trilinos, and that the seg-faulting case may occur within Trilinos, which may have been an incorrect assumption.

**Comment by ndellingwood** [comment]
If no one is using this outside of Trilinos then we should consider moving to the impl namespace and document for developers use-cases to avoid.

**Comment by mhoemmen** [comment]
@ndellingwood wrote:
> The tougher check is when a pointer to device data is passed to the function and `findRelOffset` is called from the host, which occurs in the unit test and caused the seg fault with Cuda (the input indices only needs to have `operator[]` implemented).

It sounds like the unit test was broken; we can fix that.  Do you have evidence to suggest that the function itself is broken?

**Comment by ndellingwood** [comment]
@mhoemmen 

> Do you have evidence to suggest that the function itself is broken?

The function isn't broken if used correctly AFAIK (that is works fine when not calling from host with device data without UVM enabled). Since there is no catch for possible misuse aside from a seg-fault I marked this as a bug, but I'll change the labeling if you or others disagree. The unit test as-is is broken, but pointed out there is possibility for misuse of the function. We can add documentation about how the function could be misused. If only Trilinos makes use of this I'd be in favor of moving to an impl namespace, then we don't have to worry about non-developers misusing.

**Comment by ndellingwood** [comment]
@mhoemmen To resolve the issue I'll add a PR to fix the unit test for the Cuda case and add a comment in the header about potential for misuse that should be avoided.

**Comment by jjwilke** [comment]
@ndellingwood What is the fix? I pushed a quick fix into the cmake-rebased branch to get the tests to pass.

**Comment by ndellingwood** [comment]
@jjwilke Thanks. The unit test is broken, I'll put in a separate PR for that. The function isn't broken but has potential for misuse without any mechanism for catching it, but for now I'll just add documentation.

**Comment by mhoemmen** [comment]
> Since there is no catch for possible misuse aside from a seg-fault I marked this as a bug, but I'll change the labeling if you or others disagree.

I think this is a bug in the unit test, not the function.  If any function is marked `KOKKOS_*FUNCTION` and it accesses data, then that function assumes that the data are accessible on device.  Any function that takes an array or `Kokkos::View` has that behavior.
